protocol/contracts/nexus/Relic.sol

```diff
@@ -0,0 +1,344 @@
+}
+
+contract Relic is
+    ERC721,
+    ERC721Enumerable,
+    ERC721URIStorage,
+    Pausable,
+    Ownable,
+    ERC721Burnable,
+    IERC1155Receiver,
+    ReentrancyGuard
+{
+    constructor() ERC721("Relic", "RELIC") {}
+
+    using Strings for uint256;
+    using Counters for Counters.Counter;
+    Counters.Counter public _tokenIdCounter;
+
+    enum Enclave {
+        Chaos,
+        Mystery,
+        Logic,
+        Order,
+        Structure
+    }
+    mapping(uint256 => Enclave) enclaves;
+    enum Rarity {
+        Common,
+        Uncommon,
+        Rare,
+        Epic,
+        Legendary
+    }
+    // users whitelisted for minting a Relic
+    mapping(address => bool) public whitelisted;
+    // contracts authorised to mint Relics out
+    mapping(address => bool) public whitelistedContracts;
+    // users' Relic Shard balances RELIC_ID => (SHARD_ID => BALANCE)
+    mapping(uint256 => mapping(uint256 => uint256)) public balances;
+    // relic experience points
+    mapping(uint256 => uint256) public relicXP;
+
+    // URIs served depending on relic rarity
+    mapping(Rarity => string) private BASE_URIS;
+    // shard contract interface
+    IShards public SHARDS;
+    // contract whitelisting users to mint a Relic
+    address public whitelisterAddress;
+    // XP thresholds for Relic levels
+    uint256[] public thresholds;
+
+    event MintRelic(address to, uint256 enclave);
+    event GivePoints(uint256 relicId, uint256 points);
+
+    //------- External -------//
+    // called by whitelisted users to mint Relics
+    function mintRelic(Enclave _selectedEnclave) external nonReentrant {
+        require(whitelisted[msg.sender], "You cannot own a Relic yet");
+        uint256 tokenId = _tokenIdCounter.current();
+        enclaves[tokenId] = _selectedEnclave;
+        _tokenIdCounter.increment();
+        _safeMint(msg.sender, tokenId);
+
+        whitelisted[msg.sender] = false;
+
+        emit MintRelic(msg.sender, uint256(_selectedEnclave));
+    }
+
+    // Allows users to equip Shards into their Relics
+    function batchEquipShard(
+        uint256 _targetRelic,
+        uint256[] memory _itemIds,
+        uint256[] memory _amounts
+    ) external nonReentrant {
+        require(
+            msg.sender == ownerOf(_targetRelic),
+            "You don't own this Relic"
+        );
+
+        SHARDS.equipShard(msg.sender, _itemIds, _amounts);
+
+        for (uint256 i = 0; i < _itemIds.length; i++) {
+            balances[_targetRelic][_itemIds[i]] += _amounts[i];
+        }
+    }
+    // Allows users to unequip Shards from their Relic
+    function batchUnequipShard(
+        uint256 _targetRelic,
+        uint256[] memory _itemIds,
+        uint256[] memory _amounts
+    ) external nonReentrant {
+        require(
+            msg.sender == ownerOf(_targetRelic),
+            "You don't own this Relic"
+        );
+
+        SHARDS.unEquipShard(msg.sender, _itemIds, _amounts);
```

Comment:
If the Relic contract has a bunch of shards on it for many different relic ids (ie many users), what's stopping a user unequipping any shard which it didn't equip in the first place?

Are you just relying on the fact that the below balances check reverts since you can't have a negative balance? Or some other check? Do you have a test for this?

Highly recommend using the checks-effects-interactions pattern here and anywhere else you need to transfer the NFT or ERC1155. Move any external contract interaction to the very end.